src/function/aggregate/distributive/bitstring_agg.cpp

```diff
@@ -0,0 +1,256 @@
+		return max - min + 1;
+	}
+
+	template <class INPUT_TYPE, class STATE>
+	static void Execute(STATE *state, INPUT_TYPE input, INPUT_TYPE min) {
+		Bit::SetBit(state->value, input - min, 1);
+	}
+
+	template <class STATE, class OP>
+	static void Combine(const STATE &source, STATE *target, AggregateInputData &) {
+		if (!source.is_set) {
+			return;
+		}
+		if (!target->is_set) {
+			Assign(target, source.value);
+			target->is_set = true;
+			target->min = source.min;
+			target->max = source.max;
+		} else {
+			Bit::BitwiseOr(source.value, target->value, target->value);
+		}
+	}
+
+	template <class INPUT_TYPE, class STATE>
+	static void Assign(STATE *state, INPUT_TYPE input) {
+		D_ASSERT(state->is_set == false);
+		if (input.IsInlined()) {
+			state->value = input;
+		} else { // non-inlined string, need to allocate space for it
+			auto len = input.GetSize();
+			auto ptr = new char[len];
+			memcpy(ptr, input.GetDataUnsafe(), len);
+			state->value = string_t(ptr, len);
+		}
+	}
+
+	template <class T, class STATE>
+	static void Finalize(Vector &result, AggregateInputData &, STATE *state, T *target, ValidityMask &mask, idx_t idx) {
+		if (!state->is_set) {
+			mask.SetInvalid(idx);
+		} else {
+			target[idx] = StringVector::AddStringOrBlob(result, state->value);
+		}
+	}
+
+	template <class STATE>
+	static void Destroy(STATE *state) {
+		if (state->is_set && !state->value.IsInlined()) {
+			delete[] state->value.GetDataUnsafe();
+		}
+	}
+
+	static bool IgnoreNull() {
+		return true;
+	}
+};
+
+template <>
+void BitStringAggOperation::Execute(BitAggState<hugeint_t> *state, hugeint_t input, hugeint_t min) {
+	idx_t val;
+	if (Hugeint::TryCast(input - min, val)) {
+		Bit::SetBit(state->value, val, 1);
+	} else {
+		throw OutOfRangeException("Range too large for bitstring aggregation");
+	}
+}
+
+template <>
+idx_t BitStringAggOperation::GetRange(hugeint_t min, hugeint_t max) {
+	idx_t val;
+	if (Hugeint::TryCast(max - min + 1, val)) {
+		return val;
+	} else {
+		throw OutOfRangeException("Range too large for bitstring aggregation");
+	}
+}
+
+unique_ptr<BaseStatistics> BitstringPropagateStats(ClientContext &context, BoundAggregateExpression &expr,
+                                                   FunctionData *bind_data,
+                                                   vector<unique_ptr<BaseStatistics>> &child_stats,
+                                                   NodeStatistics *node_stats) {
+
+	if (child_stats[0] && node_stats && node_stats->has_max_cardinality) {
+		auto &numeric_stats = (NumericStatistics &)*child_stats[0];
+		if (numeric_stats.min.IsNull() || numeric_stats.max.IsNull()) {
+			return nullptr;
+		}
+		auto bind_agg_data = (BitstringAggBindData *)bind_data;
+		bind_agg_data->min = numeric_stats.min;
+		bind_agg_data->max = numeric_stats.max;
+	} else {
+		throw BinderException("Could not retrieve required statistics. Alternatively, try by providing the statistics "
```

Comment:
As this exception is thrown during the propagate stats - I wonder how this works if the optimizer is disabled, as this method is not called in that case. Could we add some tests with `PRAGMA disable_optimizer`?

Reply:
I've added tests for this and check if min/max values are actually set later on in the aggregation logic instead of only catching this when setting the statistics!